File: sdk/prismer-cloud/runtime/src/cli/ui.ts
// CLI UI primitives — ported from pre-refactor `runtime/src/cli/ui.ts`.
// Provides a 6-level output hierarchy (header / banner / primary / secondary /
// tip / status indicators / error block), 3 modes (pretty / quiet / json),
// table / spinner / progress / json helpers, and brand-voice guard.
//
// MUST sync with sdk/prismer-cloud/typescript/src/cli-ui.ts
//
// This file is the canonical TUI style surface. The SDK CLI (@prismer/sdk,
// bin `cloud`) maintains a mirror at `src/cli-ui.ts` so users switching
// between `prismer ...` (daemon ops) and `cloud ...` (agent/user ops) see
// the same icons, colors, banner, table layout, spinner, and brand voice.
// Any change here must be ported to the SDK mirror in the same commit.

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

// ============================================================
// Types
// ============================================================

export type OutputMode = 'pretty' | 'json' | 'quiet';

export interface UIOptions {
  mode?: OutputMode;
  color?: boolean;
  stream?: NodeJS.WritableStream;
  errStream?: NodeJS.WritableStream;
}

export interface TableRow {
  [column: string]: string;
}

export interface TableOptions {
  columns: string[];
  maxWidth?: number;
}

export interface LegacyTableOptions {
  columns: string[];
  rows: TableRow[];
  maxWidth?: number;
}

// ============================================================
// Brand voice guard
// ============================================================

const FORBIDDEN_SUBSTRINGS = ['Sorry', 'Unfortunately', 'Oops'];
const FORBIDDEN_WORDS = ['Please'];

export function assertBrandVoice(text: string, label = 'text'): void {
  if (!process.env['PRISMER_BRAND_VOICE_STRICT']) return;
  for (const sub of FORBIDDEN_SUBSTRINGS) {
    if (text.includes(sub)) {
      throw new Error(`[Brand Voice] Forbidden substring "${sub}" in CLI ${label}: ${text}`);
    }
  }
  for (const word of FORBIDDEN_WORDS) {
    const re = new RegExp(`\\b${word}\\b`);
    if (re.test(text)) {
      throw new Error(`[Brand Voice] Forbidden word "${word}" in CLI ${label}: ${text}`);
    }
  }
  const lines = text.split('\n');
  for (const line of lines) {
    const stripped = line.replace(/"[^"]*"/g, '').replace(/'[^']*'/g, '');
    if (stripped.trimEnd().endsWith('!')) {
      throw new Error(`[Brand Voice] Trailing "!" in CLI ${label}: ${line}`);
    }
  }
}

// ============================================================
// Spinner / banner constants
// ============================================================

const BRAILLE_FRAMES = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
const COMPACT_BANNER = ['◇ PRISMER', '  Runtime CLI'];

// ============================================================
// Icon asset resolution
// ============================================================

function thisDirname(): string {
  // ESM-safe: derive __dirname from import.meta.url. For tsup CJS emits the
  // built file shadows this with a synthesized __dirname, which still resolves.
  try {
    return path.dirname(fileURLToPath(import.meta.url));
  } catch {
    return process.cwd();
  }
}

function findIconPath(size: 'big' | 'small' = 'big'): string | null {
  const name = size === 'big' ? 'icon' : 'smallicon';
  const here = thisDirname();
  const candidates = [
    // npm-installed: node_modules/@prismer/runtime/dist/cli.js → ../assets
    path.resolve(here, '../assets', name),
    // alternate dist layout (sub-bundle): dist/bin/cli.js → ../../assets
    path.resolve(here, '../../assets', name),
    // source/typecheck: src/cli/ui.ts → ../../assets
    path.resolve(here, '../../assets', name),
    // dev mode: cwd happens to be runtime root
    path.resolve(process.cwd(), 'assets', name),
    path.resolve(process.cwd(), 'sdk/prismer-cloud/runtime/assets', name),
  ];
  for (const candidate of candidates) {
    try {
      if (fs.existsSync(candidate)) return candidate;
    } catch {
      /* ignore */
    }
  }
  return null;
}

// ============================================================
// UI class
// ============================================================

export class UI {
  readonly mode: OutputMode;
  readonly colorEnabled: boolean;
  private readonly stream: NodeJS.WritableStream;
  private readonly errStream: NodeJS.WritableStream;

  constructor(opts?: UIOptions) {
    this.mode = opts?.mode ?? 'pretty';
    this.stream = opts?.stream ?? process.stdout;
    this.errStream = opts?.errStream ?? process.stderr;

    if (opts?.color !== undefined) {
      this.colorEnabled = opts.color;
    } else {
      const isTTY = (this.stream as NodeJS.WriteStream).isTTY === true;
      const noColor = Boolean(process.env['NO_COLOR']);
      this.colorEnabled = isTTY && !noColor;
    }
  }

  // ---- Internal color helpers ----

  private ansi(open: number, close: number, text: string): string {
    if (!this.colorEnabled) return text;
    return `[${open}m${text}[${close}m`;
  }

  private green(t: string): string {
    return this.ansi(32, 39, t);
  }
  private red(t: string): string {
    return this.ansi(31, 39, t);
  }
  private yellow(t: string): string {
    return this.ansi(33, 39, t);
  }
  private cyan(t: string): string {
    return this.ansi(36, 39, t);
  }
  private dim(t: string): string {
    return this.ansi(2, 22, t);
  }
  private bold(t: string): string {
    return this.ansi(1, 22, t);
  }
  private gray(t: string): string {
    return this.ansi(90, 39, t);
  }
  private brandMark(): string {
    return this.cyan('◇');
  }

  private colorBrandLine(line: string): string {
    let out = '';
    for (const ch of line) {
      if (ch === '▒') {
        out += this.cyan(ch);
      } else if (ch === '▓') {
        out += this.dim(ch);
      } else {
        out += ch;
      }
    }
    return out;
  }

  // ---- Core write helpers ----

  write(text: string): void {
    this.stream.write(text);
  }

  writeErr(text: string): void {
    this.errStream.write(text);
  }

  // ---- Level 1: Header ----

  header(text: string): void {
    if (this.mode === 'json') return;
    const prefix = text.startsWith('Prismer') ? this.brandMark() + ' ' : '';
    this.write(prefix + this.bold(text) + '\n');
  }

  smallHeader(subtitle?: string): void {
    if (this.mode === 'json' || this.mode === 'quiet') return;
    const iconPath = findIconPath('small');
    if (iconPath !== null) {
      try {
        const raw = fs.readFileSync(iconPath, 'utf-8').replace(/\n+$/, '');
        for (const line of raw.split('\n')) {
          this.write(this.cyan(line) + '\n');
        }
      } catch {
        this.write(this.brandMark() + ' ' + this.bold('Prismer') + '\n');
      }
    } else {
      this.write(this.brandMark() + ' ' + this.bold('Prismer') + '\n');
    }
    if (subtitle !== undefined && subtitle.length > 0) {
      this.write(this.dim('  ' + subtitle) + '\n');
    }
    this.blank();
  }

  banner(subtitle?: string, opts?: { full?: boolean }): void {
    if (this.mode === 'json' || this.mode === 'quiet') return;

    const envColumns =
      process.env['COLUMNS'] !== undefined ? parseInt(process.env['COLUMNS'], 10) : NaN;
    const width =
      (this.stream as NodeJS.WriteStream).columns ??
      process.stdout.columns ??
      (Number.isFinite(envColumns) ? envColumns : 80);
    const iconPath = findIconPath('big');
    const shouldUseFull = opts?.full === true || width >= 120;

    if (shouldUseFull && iconPath !== null) {
      try {
        const raw = fs.readFileSync(iconPath, 'utf-8');
        const lines = raw.split('\n');
        for (const line of lines) {
          const brandedLine = line.replace('Prismer Cloud SDK', 'Prismer Runtime CLI');
          const stripped = brandedLine.trimEnd();
          if (stripped.length === 0) {
            this.write('\n');
            continue;
          }
          const clipped =
            stripped.length >= width ? stripped.slice(0, Math.max(width - 1, 1)) : stripped;
          this.write(this.colorBrandLine(clipped) + '\n');
        }
      } catch {
        this.write(this.cyan(COMPACT_BANNER[0] ?? '◇ PRISMER') + '\n');
        this.write(this.dim(COMPACT_BANNER[1] ?? '  Runtime CLI') + '\n');
      }
    } else {
      this.write(this.cyan(COMPACT_BANNER[0] ?? '◇ PRISMER') + '\n');
      this.write(this.dim(COMPACT_BANNER[1] ?? '  Runtime CLI') + '\n');
    }

    if (subtitle !== undefined && subtitle.length > 0) {
      this.write(this.dim('  ' + subtitle) + '\n');
    }
    this.blank();
  }

  // ---- Level 2: Primary data ----

  blank(): void {
    if (this.mode === 'json') return;
    this.write('\n');
  }

  line(text: string): void {
    if (this.mode === 'json') return;
    this.write(text + '\n');
  }

  info(text: string): void {
    this.line(text);
  }

  // ---- Level 3: Secondary ----

  secondary(text: string, indent = 2): void {
    if (this.mode === 'json') return;
    this.write(' '.repeat(indent) + this.dim(text) + '\n');
  }

  // ---- Level 4: Action tips ----

  tip(text: string): void {
    if (this.mode === 'json') return;
    this.write(this.cyan('Tip:') + ' ' + text + '\n');
  }

  next(text: string): void {
    if (this.mode === 'json') return;
    this.write(this.cyan('Next:') + ' ' + text + '\n');
  }

  // ---- Level 5: Status indicators ----

  ok(text: string, detail?: string): void {
    if (this.mode === 'json') return;
    const suffix = detail ? '  ' + this.dim(detail) : '';
    this.write('  ' + this.green('✓') + ' ' + text + suffix + '\n');
  }

  success(text: string, detail?: string): void {
    this.ok(text, detail);
  }

  fail(text: string, detail?: string): void {
    if (this.mode === 'json') return;
    const suffix = detail ? '  ' + this.dim(detail) : '';
    this.write('  ' + this.red('✗') + ' ' + text + suffix + '\n');
  }

  online(text: string): void {
    if (this.mode === 'json') return;
    this.write('  ' + this.green('●') + ' ' + text + '\n');
  }

  offline(text: string): void {
    if (this.mode === 'json') return;
    this.write('  ' + this.gray('○') + ' ' + text + '\n');
  }

  notInstalled(text: string): void {
    if (this.mode === 'json') return;
    this.write('  ' + this.dim('·') + ' ' + this.dim(text) + '\n');
  }

  pending(text: string): void {
    if (this.mode === 'json') return;
    this.write('  ' + this.yellow('⟳') + ' ' + text + '\n');
  }

  warn(text: string, detail?: string): void {
    if (this.mode === 'json') return;
    const suffix = detail ? '  ' + this.dim(detail) : '';
    this.write('  ' + this.yellow('!') + ' ' + text + suffix + '\n');
  }

  // ---- Level 6: Error block ----

  error(what: string, cause?: string, fix?: string): void {
    if (this.mode === 'json') return;
    this.writeErr(this.red('✗') + ' ' + what + '\n');
    if (cause !== undefined) {
      this.writeErr('  ' + this.dim('Cause:') + ' ' + this.dim(cause) + '\n');
    }
    if (fix !== undefined) {
      this.writeErr('  ' + this.cyan('Fix:') + ' ' + fix + '\n');
    }
  }

  // ---- Tables ----

  table(rows: TableRow[], opts: TableOptions): void;
  table(opts: LegacyTableOptions): void;
  table(rowsOrOpts: TableRow[] | LegacyTableOptions, maybeOpts?: TableOptions): void {
    if (this.mode === 'json') return;
    const rows = Array.isArray(rowsOrOpts) ? rowsOrOpts : rowsOrOpts.rows;
    const opts = Array.isArray(rowsOrOpts)
      ? maybeOpts
      : { columns: rowsOrOpts.columns, maxWidth: rowsOrOpts.maxWidth };

    if (!opts) throw new Error('table() requires columns');

    const maxWidth = opts.maxWidth ?? (process.stdout.columns || 80);
    const cols = opts.columns;

    const widths: number[] = cols.map((col) => col.length);
    for (const row of rows) {
      cols.forEach((col, i) => {
        const val = row[col] ?? '';
        const w = widths[i] ?? 0;
        if (val.length > w) widths[i] = val.length;
      });
    }

    const totalWidth = widths.reduce((a, b) => a + b, 0) + (cols.length - 1) * 2 + 2;

    if (totalWidth > maxWidth) {
      // List mode
      for (let i = 0; i < rows.length; i++) {
        const row = rows[i];
        if (!row) continue;
        for (const col of cols) {
          const val = row[col] ?? '';
          this.write('  ' + this.bold(col + ':') + ' ' + val + '\n');
        }
        if (i < rows.length - 1) this.write('\n');
      }
      return;
    }

    const header = cols.map((col, i) => col.toUpperCase().padEnd(widths[i] ?? col.length)).join('  ');
    this.write('  ' + this.dim(header) + '\n');

    for (const row of rows) {
      const line = cols.map((col, i) => (row[col] ?? '').padEnd(widths[i] ?? col.length)).join('  ');
      this.write('  ' + line + '\n');
    }
  }

  // ---- Spinner ----

  spinner(text: string): { update(t: string): void; stop(final?: string): void } {
    if (this.mode === 'quiet' || this.mode === 'json') {
      return {
        update(): void {
          /* no-op */
        },
        stop(): void {
          /* no-op */
        },
      };
    }

    const isTTY = (this.stream as NodeJS.WriteStream).isTTY === true;

    if (!isTTY || !this.colorEnabled) {
      this.write('  ' + this.yellow('⟳') + ' ' + text + '\n');
      return {
        update: (t: string) => {
          this.write('  ' + this.yellow('⟳') + ' ' + t + '\n');
        },
        stop: (final?: string) => {
          if (final) this.write('  ' + this.green('✓') + ' ' + final + '\n');
        },
      };
    }

    let current = text;
    let frameIdx = 0;
    let stopped = false;

    const write = this.write.bind(this);
    const colorFn = this.yellow.bind(this);
    const greenFn = this.green.bind(this);

    function renderFrame(): void {
      const frame = BRAILLE_FRAMES[frameIdx % BRAILLE_FRAMES.length] ?? '⠋';
      const line = '  ' + colorFn(frame) + ' ' + current;
      write('\r' + line);
      frameIdx++;
    }

    renderFrame();
    const timer = setInterval(renderFrame, 80);

    return {
      update(t: string): void {
        if (stopped) return;
        current = t;
      },
      stop(final?: string): void {
        if (stopped) return;
        stopped = true;
        clearInterval(timer);
        write('\r\x1b[2K');
        if (final) write('  ' + greenFn('✓') + ' ' + final + '\n');
      },
    };
  }

  // ---- Progress bar ----

  progress(
    text: string,
    total: number,
  ): { update(current: number, detail?: string): void; stop(final?: string): void } {
    if (this.mode === 'quiet' || this.mode === 'json') {
      return {
        update(): void {
          /* no-op */
        },
        stop(): void {
          /* no-op */
        },
      };
    }

    const isTTY = (this.stream as NodeJS.WriteStream).isTTY === true;
    const start = Date.now();
    const write = this.write.bind(this);
    const colorFn = this.cyan.bind(this);
    const dimFn = this.dim.bind(this);
    const greenFn = this.green.bind(this);

    let last = 0;
    let lastDetail = '';
    let stopped = false;

    const render = (): void => {
      if (stopped) return;
      const frac = total > 0 ? Math.min(1, Math.max(0, last / total)) : 0;
      const pct = Math.floor(frac * 100);
      const width = 20;
      const filled = Math.floor(frac * width);
      const bar = '█'.repeat(filled) + '░'.repeat(width - filled);
      const elapsed = (Date.now() - start) / 1000;
      const eta = frac > 0.01 ? Math.max(0, elapsed / frac - elapsed) : 0;
      const etaStr = frac >= 1 ? '' : ` · ${eta < 1 ? '<1s' : Math.round(eta) + 's'} left`;
      const detailStr = lastDetail ? ` · ${lastDetail}` : '';
      const line = `  ${text} [${colorFn(bar)}] ${String(pct).padStart(3)}%${detailStr}${dimFn(etaStr)}`;
      if (isTTY && this.colorEnabled) {
        write('\r\x1b[2K' + line);
      } else {
        write(line + '\n');
      }
    };

    render();

    return {
      update: (current: number, detail?: string): void => {
        if (stopped) return;
        last = current;
        if (detail !== undefined) lastDetail = detail;
        render();
      },
      stop: (final?: string): void => {
        if (stopped) return;
        stopped = true;
        if (isTTY && this.colorEnabled) write('\r\x1b[2K');
        if (final) write('  ' + greenFn('✓') + ' ' + final + '\n');
      },
    };
  }

  // ---- JSON output ----

  json(payload: unknown, opts?: { pretty?: boolean }): void {
    const indent = opts?.pretty ? 2 : undefined;
    this.write(JSON.stringify(payload, null, indent) + '\n');
  }

  result<T>(pretty: () => void, jsonPayload: T): void {
    if (this.mode === 'pretty') {
      pretty();
    } else {
      this.json(jsonPayload);
    }
  }
}

// ============================================================
// Singleton
// ============================================================

let _ui: UI | null = null;

export function getUI(): UI {
  if (!_ui) _ui = new UI();
  return _ui;
}

export function setUI(ui: UI): void {
  _ui = ui;
}

export function __resetUIForTests(): void {
  _ui = null;
}

// ============================================================
// applyCommonFlags — strips --json/--quiet/--no-color/--color from argv
// ============================================================

export function applyCommonFlags(argv: string[]): {
  mode: OutputMode;
  color: boolean;
  restArgv: string[];
} {
  let mode: OutputMode = 'pretty';
  const isTTY = process.stdout.isTTY === true;
  const noColorEnv = Boolean(process.env['NO_COLOR']);
  let color = isTTY && !noColorEnv;

  const rest: string[] = [];

  for (const arg of argv) {
    switch (arg) {
      case '--no-color':
        color = false;
        break;
      case '--color':
        color = true;
        break;
      case '--json':
      case '--pretty-json':
        mode = 'json';
        // Push --json back so commander parses it per-command too — every
        // subcommand declares `.option('--json', ...)`, so leaving it in argv
        // makes opts.json truthy in action handlers (used to gate JSON output).
        if (arg === '--json') rest.push(arg);
        break;
      case '--quiet':
        mode = 'quiet';
        break;
      default:
        rest.push(arg);
    }
  }

  return { mode, color, restArgv: rest };
}

File: sdk/prismer-cloud/rust/src/ui.rs
//! Simplified CLI UI primitives per §15 cli-design spec
//!
//! This module provides consistent, beautiful CLI output across Prismer SDKs.
//! Design reference: docs/version190/15-cli-design.md

use std::env;

// ============================================================================
// Color codes
// ============================================================================

pub const COLOR_RESET: &str = "\x1b[0m";
pub const COLOR_RED: &str = "\x1b[31m";
pub const COLOR_GREEN: &str = "\x1b[32m";
pub const COLOR_YELLOW: &str = "\x1b[33m";
pub const COLOR_CYAN: &str = "\x1b[36m";
pub const COLOR_DIM: &str = "\x1b[2m";
pub const COLOR_BOLD: &str = "\x1b[1m";

// ============================================================================
// Status indicators per §15
// ============================================================================

pub const STATUS_OK: &str = "✓";
pub const STATUS_FAIL: &str = "✗";
pub const STATUS_ONLINE: &str = "●";
pub const STATUS_OFFLINE: &str = "○";
pub const STATUS_PENDING: &str = "⟳";
pub const STATUS_NOT_INSTALLED: &str = "·";

// ============================================================================
// Brand icon - compact mode only
// ============================================================================

pub const COMPACT_BANNER: [&str; 2] = ["◇ PRISMER", "  Runtime CLI"];

// ============================================================================
// Output mode
// ============================================================================

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OutputMode {
    Pretty,
    JSON,
    Quiet,
}

// ============================================================================
// UI struct
// ============================================================================

pub struct UI {
    color_enabled: bool,
    mode: OutputMode,
    width: usize,
}

impl UI {
    pub fn new() -> Self {
        let color_enabled = supports_color();
        let width = get_terminal_width();
        UI {
            color_enabled,
            mode: OutputMode::Pretty,
            width,
        }
    }

    pub fn set_mode(&mut self, mode: OutputMode) {
        self.mode = mode;
    }

    pub fn set_color(&mut self, enabled: bool) {
        self.color_enabled = enabled;
    }

    fn ansi(&self, code: &str, text: &str) -> String {
        if !self.color_enabled {
            return text.to_string();
        }
        format!("{}{}{}", code, text, COLOR_RESET)
    }

    fn red(&self, text: &str) -> String {
        self.ansi(COLOR_RED, text)
    }

    fn green(&self, text: &str) -> String {
        self.ansi(COLOR_GREEN, text)
    }

    fn yellow(&self, text: &str) -> String {
        self.ansi(COLOR_YELLOW, text)
    }

    fn cyan(&self, text: &str) -> String {
        self.ansi(COLOR_CYAN, text)
    }

    fn dim(&self, text: &str) -> String {
        self.ansi(COLOR_DIM, text)
    }

    fn bold(&self, text: &str) -> String {
        self.ansi(COLOR_BOLD, text)
    }

    // Level 1: Header
    pub fn header(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("{}", self.bold(text));
    }

    pub fn banner(&self, subtitle: Option<&str>, full: bool) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }

        if full && self.width >= 120 {
            for line in COMPACT_BANNER {
                println!("{}", self.cyan(line));
            }
        } else {
            println!("{}", self.cyan("◇ PRISMER"));
            println!("{}", self.dim("  Runtime CLI"));
        }

        if let Some(sub) = subtitle {
            println!("  {}", self.dim(sub));
        }
        println!();
    }

    // Level 2: Primary data
    pub fn blank(&self) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!();
    }

    pub fn line(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("{}", text);
    }

    pub fn info(&self, text: &str) {
        self.line(text);
    }

    // Level 3: Secondary
    pub fn secondary(&self, text: &str, indent: usize) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        let _padding: String = " ".repeat(indent);
        println!("{}{}", _padding, self.dim(text));
    }

    // Level 4: Action tips
    pub fn tip(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("Tip: {}", self.cyan(text));
    }

    pub fn next(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("Next: {}", self.cyan(text));
    }

    // Level 5: Status indicators
    pub fn ok(&self, text: &str, detail: Option<&str>) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        let suffix = match detail {
            Some(d) => format!("  {}", self.dim(d)),
            None => String::new(),
        };
        println!("  {} {}{}", self.green(STATUS_OK), text, suffix);
    }

    pub fn success(&self, text: &str, detail: Option<&str>) {
        self.ok(text, detail);
    }

    pub fn fail(&self, text: &str, detail: Option<&str>) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        let suffix = match detail {
            Some(d) => format!("  {}", self.dim(d)),
            None => String::new(),
        };
        println!("  {} {}{}", self.red(STATUS_FAIL), text, suffix);
    }

    pub fn online(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("  {} {}", self.green(STATUS_ONLINE), text);
    }

    pub fn offline(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("  {} {}", self.dim(STATUS_OFFLINE), text);
    }

    pub fn not_installed(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("  {} {}", self.dim(STATUS_NOT_INSTALLED), self.dim(text));
    }

    pub fn pending(&self, text: &str) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        println!("  {} {}", self.yellow(STATUS_PENDING), text);
    }

    pub fn warn(&self, text: &str, detail: Option<&str>) {
        if self.mode == OutputMode::Quiet || self.mode == OutputMode::JSON {
            return;
        }
        let suffix = match detail {
            Some(d) => format!("  {}", self.dim(d)),
            None => String::new(),
        };
        println!("  ! {}{}", text, suffix);
    }

    // Level 6: Error block
    pub fn error(&self, what: &str, cause: Option<&str>, fix: Option<&str>) {
        if self.mode == OutputMode::JSON {
            self.write_json_error(what, cause, fix);
            return;
        }
        eprintln!("{} {}", self.red(STATUS_FAIL), what);
        if let Some(c) = cause {
            eprintln!("  Cause: {}", self.dim(c));
        }
        if let Some(f) = fix {
            eprintln!("  Fix: {}{}", self.cyan(f), COLOR_RESET);
        }
    }

    fn write_json_error(&self, what: &str, cause: Option<&str>, fix: Option<&str>) {
        use serde_json::{json, Value};
        let mut err = json!({"error": what});
        if let Some(c) = cause {
            err["cause"] = Value::String(c.to_string());
        }
        if let Some(f) = fix {
            err["fix"] = Value::String(f.to_string());
        }
        eprintln!("{}", serde_json::to_string_pretty(&err).unwrap_or_default());
    }

    // JSON output
    pub fn json(&self, payload: &serde_json::Value) {
        let output = if payload.is_object() {
            serde_json::to_string_pretty(payload).unwrap_or_default()
        } else {
            payload.to_string()
        };
        println!("{}", output);
    }

    pub fn result<T: serde::Serialize>(&self, pretty: impl FnOnce(), json_payload: T) {
        match self.mode {
            OutputMode::Pretty => pretty(),
            _ => self.json(&serde_json::to_value(json_payload).unwrap_or(serde_json::json!(null))),
        }
    }

    // Brand voice checker
    pub fn check_brand_voice(text: &str, field: &str) -> Result<(), String> {
        // Check for forbidden substrings
        for fs in ["Sorry", "Unfortunately", "Oops"].iter() {
            if text.contains(fs) {
                return Err(format!("Brand voice error in {}: found forbidden substring {:?}", field, fs));
            }
        }

        // Check for forbidden words (whole word match)
        for fw in ["Please"].iter() {
            let words: Vec<&str> = text.split_whitespace().collect();
            if words.iter().any(|w| *w == *fw) {
                return Err(format!("Brand voice error in {}: found forbidden word {:?}", field, fw));
            }
        }

        // Check for trailing exclamation marks
        for line in text.lines() {
            let trimmed = line.trim();
            if trimmed.is_empty() {
                continue;
            }

            let in_double_quote = trimmed.contains('"');
            let in_single_quote = trimmed.contains('\'');
            if in_double_quote || in_single_quote {
                continue;
            }

            if trimmed.ends_with('!') && !trimmed.ends_with("!!") {
                return Err(format!("Brand voice error in {}: trailing exclamation mark in {:?}", field, line));
            }
        }

        Ok(())
    }
}

pub fn get_terminal_width() -> usize {
    if let Ok(cols) = env::var("COLUMNS") {
        if let Ok(width) = cols.parse::<usize>() {
            if width > 0 {
                return width;
            }
        }
    }
    80
}

fn supports_color() -> bool {
    env::var("NO_COLOR").is_ok()
}
